core/include/kernel/notif.h

```diff
@@ -124,6 +129,13 @@ static inline void notif_send_async(uint32_t value __unused)
  */
 TEE_Result notif_send_sync(uint32_t value);
 
+/*
+ * Send a synchronous value with timeout, note that it must be <=
+ * NOTIF_VALUE_MAX. The notification is synchronous even if the value
+ * happens to belong in the asynchronous range.
+ */
+TEE_Result notif_send_sync_timeout(uint32_t value, uint32_t tmo);
```

Comment:
Why is this function needed? What does it do that `notif_send_sync()` doesn't?

Reply:
Perhaps we can simply use `notif_send_sync()` with tmo(timeout).

Reply:
The only purpose of this function is to match the function pointer argument for `do_notif()` in `wq_wake_next()`. That is better done with an internal wrapper function just above `wq_wake_next()` that discards the timeout argument and calls `notif_send_sync()`.

Reply:
Ok, I will try to modify.

Reply:
I'm not quite sure what ```discards the timeout argument```  means. We have changed the number of parameters for ```do_notify()```, so the timeout argument is necessary, right?
In my opinion, we can expand the parameters of ```notif_send_sync()``` and ```notif_wait()``` by adding a timeout argument, so we don't need to add ```notif_send_sync_timeout()``` and ```notif_wait_timeout()``` separately.

Reply:
I tried what I suggested, but it got a bit more complicated with the function pointer than I liked. So I suggest doing something like this instead:
``` diff
diff --git a/core/kernel/wait_queue.c b/core/kernel/wait_queue.c
index 670e7c2ba263..373f7a2bc4b3 100644
--- a/core/kernel/wait_queue.c
+++ b/core/kernel/wait_queue.c
@@ -20,10 +20,10 @@ void wq_init(struct wait_queue *wq)
 	*wq = (struct wait_queue)WAIT_QUEUE_INITIALIZER;
 }
 
-static uint32_t do_notif(TEE_Result (*fn)(uint32_t, uint32_t), int id, uint32_t tmo,
-		     const char *cmd_str __maybe_unused,
-		     const void *sync_obj __maybe_unused,
-		     const char *fname, int lineno __maybe_unused)
+static TEE_Result do_notif(bool sleep, int id, uint32_t tmo,
+			   const char *cmd_str __maybe_unused,
+			   const void *sync_obj __maybe_unused,
+			   const char *fname, int lineno __maybe_unused)
 {
 	TEE_Result res = TEE_SUCCESS;
 
@@ -33,7 +33,10 @@ static uint32_t do_notif(TEE_Result (*fn)(uint32_t, uint32_t), int id, uint32_t
 	else
 		DMSG("%s thread %d %p", cmd_str, id, sync_obj);
 
-	res = fn(id + NOTIF_SYNC_VALUE_BASE, tmo);
+	if (sleep)
+		res = notif_wait_timeout(id + NOTIF_SYNC_VALUE_BASE, tmo);
+	else
+		res = notif_send_sync(id + NOTIF_SYNC_VALUE_BASE);
 	if (res)
 		DMSG("%s thread %d res %#"PRIx32, cmd_str, id, res);
 
@@ -81,8 +84,8 @@ static uint32_t wq_wait_final_helper(struct wait_queue *wq,
 	uint32_t ret;
 
 	do {
-		ret = do_notif(notif_wait_timeout, wqe->handle, tmo,
-			 "sleep", sync_obj, fname, lineno);
+		ret = do_notif(true /*sleep*/, wqe->handle, tmo,
+			       "sleep", sync_obj, fname, lineno);
 
 		old_itr_status = cpu_spin_lock_xsave(&wq_spin_lock);
 
@@ -153,8 +156,8 @@ void wq_wake_next(struct wait_queue *wq, const void *sync_obj,
 		cpu_spin_unlock_xrestore(&wq_spin_lock, old_itr_status);
 
 		if (do_wakeup)
-			do_notif(notif_send_sync_timeout, handle, 0,
-				 "wake ", sync_obj, fname, lineno);
+			do_notif(false /*!sleep*/, handle, 0, "wake ",
+				 sync_obj, fname, lineno);
 
 		if (!do_wakeup || !wake_read)
 			break;
```
with this, there's no need for `notif_send_sync_timeout()`

Reply:
Thank you for your suggestion. I have modified the new patch based on it.